netty/src/test/java/io/grpc/transport/netty/NettyClientTransportTest.java

```diff
@@ -0,0 +1,263 @@
+
+  private InetSocketAddress address;
+  private NettyServer server;
+  private List<NettyClientTransport> transports;
+  private List<NioEventLoopGroup> groups;
+
+  @Before
+  public void setup() throws Exception {
+    MockitoAnnotations.initMocks(this);
+    groups = new ArrayList<NioEventLoopGroup>();
+    transports = new ArrayList<NettyClientTransport>();
+
+    // Start the server.
+    address = TestUtils.testServerAddress(TestUtils.pickUnusedPort());
+    File serverCert = TestUtils.loadCert("server1.pem");
+    File key = TestUtils.loadCert("server1.key");
+    SslContext serverContext = GrpcSslContexts.forServer(serverCert, key).build();
+    server = new NettyServer(address, NioServerSocketChannel.class,
+            newGroup(), newGroup(), serverContext, 100, DEFAULT_WINDOW_SIZE, DEFAULT_WINDOW_SIZE);
+    server.start(new TestServerListener());
+  }
+
+  @After
+  public void teardown() throws Exception {
+    for (NettyClientTransport transport : transports) {
+      transport.shutdown();
+    }
+
+    if (server != null) {
+      server.shutdown();
+    }
+
+    for (NioEventLoopGroup group : groups) {
+      group.shutdownGracefully(0, 10, TimeUnit.SECONDS);
+    }
+  }
+
+  /**
+   * Verifies that we can create multiple TLS client transports from the same builder.
+   */
+  @Test
+  public void creatingMultipleTlsTransportsShouldSucceed() throws Exception {
+    // Create the protocol negotiator.
+    File clientCert = TestUtils.loadCert("ca.pem");
+    SslContext clientContext = GrpcSslContexts.forClient().trustManager(clientCert).build();
+    ProtocolNegotiator negotiator = ProtocolNegotiators.tls(clientContext, address);
+
+    // Create a couple client transports.
+    for (int index = 0; index < 2; ++index) {
+      NettyClientTransport transport = new NettyClientTransport(address, NioSocketChannel.class,
+              newGroup(), negotiator, DEFAULT_WINDOW_SIZE, DEFAULT_WINDOW_SIZE);
+      transports.add(transport);
+      transport.start(clientTransportListener);
+    }
+
+    // Send a single RPC on each transport.
+    final List<SettableFuture> rpcFutures = new ArrayList<SettableFuture>(transports.size());
+    MethodDescriptor<String, String> method = MethodDescriptor.create(MethodType.UNARY,
+            "/testService/test", 10, TimeUnit.SECONDS, StringMarshaller.INSTANCE,
+            StringMarshaller.INSTANCE);
+    for (NettyClientTransport transport : transports) {
+      SettableFuture rpcFuture = SettableFuture.create();
+      rpcFutures.add(rpcFuture);
+      ClientStream stream = transport.newStream(method, new Metadata.Headers(),
+              new TestClientStreamListener(rpcFuture));
+      stream.request(1);
+      stream.writeMessage(messageStream());
+      stream.halfClose();
+    }
+
+    // Wait for the RPCs to complete.
+    for (SettableFuture rpcFuture : rpcFutures) {
+      rpcFuture.get(10, TimeUnit.SECONDS);
+    }
+  }
+
+  private NioEventLoopGroup newGroup() {
```

Comment:
Why do you need multiple groups? Can't you just have a single one and share it between all usages?